presto-main/src/main/java/com/facebook/presto/type/khyperloglog/KHyperLogLog.java

```diff
@@ -0,0 +1,349 @@
+
+import com.facebook.airlift.stats.cardinality.HyperLogLog;
+import com.google.common.collect.Sets;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.slice.Murmur3Hash128;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+import it.unimi.dsi.fastutil.ints.IntArrayList;
+import it.unimi.dsi.fastutil.ints.IntList;
+import it.unimi.dsi.fastutil.longs.Long2DoubleMap;
+import it.unimi.dsi.fastutil.longs.Long2DoubleOpenHashMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectRBTreeMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectSortedMap;
+import it.unimi.dsi.fastutil.longs.LongIterator;
+import it.unimi.dsi.fastutil.longs.LongRBTreeSet;
+import it.unimi.dsi.fastutil.longs.LongSortedSet;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.PrimitiveIterator;
+import java.util.stream.LongStream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.slice.SizeOf.SIZE_OF_BYTE;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+import static io.airlift.slice.SizeOf.SIZE_OF_LONG;
+import static io.airlift.slice.Slices.wrappedIntArray;
+import static io.airlift.slice.Slices.wrappedLongArray;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * For reference on KHyperLogLog, see "KHyperLogLog: Estimating Reidentifiability and
+ * Joinability of Large Data at Scale" by Chia et al., 2019.
+ */
+public class KHyperLogLog
+{
+    public static final int DEFAULT_HLL_BUCKETS = 256;
+    public static final int DEFAULT_MAX_SIZE = 4096;
+    private static final byte VERSION_BYTE = 1;
+    private static final long HASH_OUTPUT_HALF_RANGE = Long.MAX_VALUE;
+    private static final int SIZE_OF_KHYPERLOGLOG = ClassLayout.parseClass(KHyperLogLog.class).instanceSize();
+    private static final int SIZE_OF_RBTREEMAP = ClassLayout.parseClass(Long2ObjectRBTreeMap.class).instanceSize();
+
+    private final Long2ObjectSortedMap<HyperLogLog> minhash;
+    private final int maxSize;
+    private final int hllBuckets;
+
+    private int hllsTotalEstimatedInMemorySize;
+    private int hllsTotalEstimatedSerializedSize;
+
+    public KHyperLogLog()
+    {
+        this(DEFAULT_MAX_SIZE, DEFAULT_HLL_BUCKETS, new Long2ObjectRBTreeMap<>());
+    }
+
+    public KHyperLogLog(int maxSize, int hllBuckets)
+    {
+        this(maxSize, hllBuckets, new Long2ObjectRBTreeMap<>());
+    }
+
+    public KHyperLogLog(int maxSize, int hllBuckets, Long2ObjectSortedMap<HyperLogLog> minhash)
+    {
+        this.maxSize = maxSize;
+        this.hllBuckets = hllBuckets;
+        this.minhash = requireNonNull(minhash, "minhash is null");
+        hllsTotalEstimatedInMemorySize = 0;
+        hllsTotalEstimatedSerializedSize = 0;
+
+        for (HyperLogLog hll : minhash.values()) {
```

Comment:
nit: how about `minhash.values().stream().forEach(this:: increaseTotalHllSize)`?

Reply:
Would `minhash.values().forEach(this::increaseTotalHllSize)` be better?

Reply:
Nice catch. Definitely :D